<?xml version='1.0'?>
<sdf version='1.7'>
  <world name="workshop">

  <light name="sun_light" type="directional">
    <pose>0.08 0.6 2 0 0 0</pose>
    <direction>-0.2 -0.5 -1</direction>
    <diffuse>0.6 0.6 0.6 1</diffuse>
    <specular>0.05 0.05 0.05 1</specular> 
    <cast_shadows>true</cast_shadows>
  </light>

  <light name="overhead_light_1" type="spot">
    <pose>0 0 3 0 0 0</pose> <!-- center of room -->
    <diffuse>0.35 0.35 0.35 1</diffuse>
    <specular>0.01 0.01 0.01 1</specular>
    <attenuation>
      <range>6</range>
      <constant>0.3</constant>
      <linear>0.01</linear>
      <quadratic>0.001</quadratic>
    </attenuation>
    <direction>0 0 -1</direction>
    <spot>
      <inner_angle>0.6</inner_angle>
      <outer_angle>1.0</outer_angle>
      <falloff>0.5</falloff>
    </spot>
    <cast_shadows>true</cast_shadows>
  </light>

  <light name="overhead_light_2" type="spot">
    <pose>-1.5 -1 3 0 0 0</pose> <!-- toward shelf -->
    <diffuse>0.5 0.5 0.5 1</diffuse> 
    <specular>0.01 0.01 0.01 1</specular>
    <attenuation>
      <range>6</range>
      <constant>0.4</constant>
      <linear>0.01</linear>
      <quadratic>0.001</quadratic>
    </attenuation>
    <direction>0 0 -1</direction>
    <spot>
      <inner_angle>0.6</inner_angle>
      <outer_angle>1.0</outer_angle>
      <falloff>0.5</falloff>
    </spot>
    <cast_shadows>false</cast_shadows>
  </light>

  <scene>
    <ambient>0.12 0.12 0.12 1</ambient> 
    <background>0.5 0.5 0.5 1</background>
  </scene>

    <!-- Floor -->
    <model name="floor">
      <static>true</static>
      <link name="floor_link">
        <collision name="floor_collision">
          <geometry>
            <box>
              <size>6 6 0.1</size>
            </box>
          </geometry>
        </collision>
        <visual name="floor_visual">
          <geometry>
            <box>
              <size>6 6 0.1</size>
            </box>
          </geometry>
        </visual>
      </link>
    </model>

    <!-- Walls -->
    <model name='Wall_left'>
      <static>true</static>
      <link name='wall_left_link'>
        <collision name='wall_left_collision'>
          <geometry>
            <box>
              <size>5 0.15 2.5</size>
            </box>
          </geometry>
        </collision>
        <visual name='wall_left_visual'>
          <geometry>
            <box>
              <size>5 0.15 2.5</size>
            </box>
          </geometry>
        </visual>
      </link>
      <pose>2.425 0 1.25 0 0 1.5708</pose>
    </model>

    <model name='Wall_back'>
      <static>true</static>
      <link name='wall_back_link'>
        <collision name='wall_back_collision'>
          <geometry>
            <box>
              <size>5 0.15 2.5</size>
            </box>
          </geometry>
        </collision>
        <visual name='wall_back_visual'>
          <geometry>
            <box>
              <size>5 0.15 2.5</size>
            </box>
          </geometry>
        </visual>
      </link>
      <pose>0 -2.425 1.25 0 0 0</pose>
    </model>

    <model name='Wall_right'>
      <static>true</static>
      <link name='wall_right_link'>
        <collision name='wall_right_collision'>
          <geometry>
            <box>
              <size>5 0.15 2.5</size>
            </box>
          </geometry>
        </collision>
        <visual name='wall_right_visual'>
          <geometry>
            <box>
              <size>5 0.15 2.5</size>
            </box>
          </geometry>
        </visual>
      </link>
      <pose>-2.425 0 1.25 0 0 1.5708</pose>
    </model>

    <!-- Shelf -->
    <include>
      <uri>model://StorageRack</uri>  
      <pose>-1.5 -1.9 0.05 0 0 0</pose>
    </include>
    
    <!-- Table -->
    <include>
      <uri>model://Table</uri>
      <pose>0 -0.5 0 0 0 0 1.57</pose>
    </include>

    <!-- Table Camera (RGB) -->
    <model name="table_camera">
      <static>true</static>
      <link name="table_camera_link">
        <sensor name="rgb_camera_table" type="camera">
          <pose>0 -2 2 0 0.5 1.57</pose>
          <update_rate>30</update_rate>
          <camera>
            <horizontal_fov>1.5</horizontal_fov>
            <image>
              <width>320</width>
              <height>240</height>
              <format>R8G8B8</format>
            </image>
            <clip>
              <near>0.1</near>
              <far>10</far>
            </clip>
          </camera>
          <plugin name="table_camera_plugin" filename="libgazebo_ros_camera.so">
            <ros>
              <namespace>/table_camera</namespace>
              <remapping>image_raw:=/table_camera/rgb_camera_table/image_raw</remapping>
            </ros>
          </plugin>
        </sensor>
      </link>
    </model>

    <!-- Shelf Camera (RGB) -->
    <model name="shelf_camera">
      <static>true</static>
      <link name="shelf_camera_link">
        <sensor name="rgb_camera_shelf" type="camera">
          <pose>-1.5 -2.3 1.45 0 0.5 1.57</pose>
          <update_rate>30</update_rate>
          <camera>
            <horizontal_fov>1.5</horizontal_fov>
            <image>
              <width>320</width>
              <height>240</height>
              <format>R8G8B8</format>
            </image>
            <clip>
              <near>0.1</near>
              <far>5.0</far>
            </clip>
          </camera>
          <plugin name="shelf_camera_plugin" filename="libgazebo_ros_camera.so">
            <ros>
              <namespace>/shelf_camera</namespace>
              <remapping>image_raw:=/shelf_camera/rgb_camera_shelf/image_raw</remapping>
            </ros>
          </plugin>
        </sensor>
      </link>
    </model>

    <!-- Room Camera -->
    <model name="room_camera">
      <static>true</static>
      <link name="room_camera_link">
        <sensor name="room_camera" type="camera">
          <pose>-0.5 1.3 2 0 0.5 -1.57</pose> 
          <camera>
            <horizontal_fov>1.3962634</horizontal_fov>
            <image>
              <width>320</width>
              <height>240</height>
              <format>R8G8B8</format>
            </image>
            <clip>
              <near>0.1</near>
              <far>100</far>
            </clip>
          </camera>
          <plugin name="room_camera_plugin" filename="libgazebo_ros_camera.so">
            <ros>
              <namespace>/room_camera</namespace>
              <remapping>image_raw:=image_raw</remapping>
              <remapping>camera_info:=camera_info</remapping>
            </ros>
            <cameraName>room_camera</cameraName>
            <frameName>room_camera_link</frameName>
          </plugin>
        </sensor>
      </link>
    </model>
  </world>
</sdf>
